src/backends/r1cs/mod.rs

```diff
@@ -0,0 +1,557 @@
+    fn add(&mut self, lhs: &CellVar, rhs: &CellVar, span: crate::constants::Span) -> CellVar {
+        let one = Fr::from(1);
+        let zero = Fr::from(0);
+
+        let res = self.new_internal_var(Value::LinearCombination(vec![(one, *lhs), (one, *rhs)], zero), span);
+
+        // IMPORTANT: since terms use CellVar as key,
+        // HashMap automatically overrides it to single one term if the two vars are the same CellVar
+        let a = if lhs == rhs {
+            LinearCombination {
+                terms: Some(HashMap::from_iter(vec![(*lhs, Fr::from(2))])),
+                constant: None,
+            }
+        } else {
+            LinearCombination {
+                terms: Some(HashMap::from_iter(vec![(*lhs, one), (*rhs, one)])),
+                constant: None,
+            }
+        };
+
+        let b = LinearCombination {
+            terms: None,
+            constant: Some(one),
+        };
+
+        let c = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(res, one)])),
+            constant: None,
+        };
+
+        self.add_constraint(
+            "add constraint: lhs + rhs = res",
+            Constraint{a, b, c},
+            span
+        );
+
+        res
+    }
+    
+    /// to constraint:
+    /// x + cst = res
+    /// given:
+    /// a * b = c
+    /// then:
+    /// a = x + cst
+    /// b = 1
+    /// c = res
+    fn add_const(&mut self, x: &CellVar, cst: &Self::Field, span: crate::constants::Span) -> CellVar {
+        let one = Fr::from(1);
+
+        let res = self.new_internal_var(Value::LinearCombination(vec![(one, *x)], *cst), span);
+
+        let a = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(*x, one)])),
+            constant: Some(*cst),
+        };
+
+        let b = LinearCombination {
+            terms: None,
+            constant: Some(one),
+        };
+
+        let c = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(res, one)])),
+            constant: None,
+        };
+
+        self.add_constraint(
+            "add constraint: x + cst = res",
+            Constraint{a, b, c},
+            span
+        );
+
+        res
+    }
+    
+    /// to constraint:
+    /// lhs * rhs = res
+    /// given:
+    /// a * b = c
+    /// then:
+    /// a = lhs
+    /// b = rhs
+    /// c = res
+    fn mul(&mut self, lhs: &CellVar, rhs: &CellVar, span: crate::constants::Span) -> CellVar {
+        let one = Fr::from(1);
+
+        let res = self.new_internal_var(Value::Mul(*lhs, *rhs), span);
+
+        let a = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(*lhs, one)])),
+            constant: None,
+        };
+
+        let b = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(*rhs, one)])),
+            constant: None,
+        };
+
+        let c = LinearCombination {
+            terms: Some(HashMap::from_iter(vec![(res, one)])),
+            constant: None,
+        };
```

Comment:
`LinearCombination::from_var(res)` would be cleaner

Reply:
implemented from_vars :D